content/extending-chat-widget/custom-identity-provider/index.mdx

```diff
@@ -0,0 +1,174 @@
+
+
+# Examples
+
+Here are our suggested implementations of Customer Identity Provider. The `fetchLiveChatToken()` method used in examples is used to acquire the [Chat Widget token](#chat-widget-token).
+
+## Without caching
+
+If you don't need to cache the token, the fetching logic shall occur on every refresh of your application page. In such a case, the implementation can be very simple - we only have to ensure handling of `getToken` promise and simply resolve others.
+
+```js
+ window.__lc.custom_identity_provider = () => {
+    const fetchLiveChatToken = () => {
+        // fetch a token from LiveChat Accounts API here
+    }
+
+    let tokenPromise = null
+    const fetchToken = () => {
+        tokenPromise = fetchLiveChatToken()
+            .then(response => {
+                tokenPromise = null
+                return response
+            })
+    }
+
+    return {
+        getToken: () => tokenPromise || fetchToken(),
+        getFreshToken: () => Promise.resolve(),
+        hasToken: () => Promise.resolve(false),
+        invalidate: () => Promise.resolve()
+    }
+ }
+```
+
+## With caching
+
+If you care about the number of requests being made in your application, you can unleash the full potential of the Custom Identity Provider.
+
+```js
+window.__lc.custom_identity_provider = () => {
+    const fetchLiveChatToken = () => {
+        // fetch a token from LiveChat Accounts API here
+    }
+
+    const cacheKey = 'YOUR_CACHE_KEY'
+    let tokenPromise = null
+    let cachedToken = null
+
+    const isExpired = ({ creationDate, expiresIn }) => Date.now() >= creationDate + expiresIn
+
+    let retrievingToken = window.sessionStorage.getItem(cacheKey).then(token => {
+        if (!retrievingToken) {
+            return
+        }
+
+        retrievingToken = null
+
+        if (!token) {
+            return
+        }
+
+        cachedToken = JSON.parse(token)
+    })
+
+    const getToken = () => {
+        if (pendingTokenRequest) {
+			return pendingTokenRequest
+		}
+
+		if (cachedToken && !isExpired(cachedToken)) {
+			return Promise.resolve(cachedToken)
+		}
+
+		if (retrievingToken) {
+			return retrievingToken.then(getToken)
+		}
+
+		return getFreshToken()
+    }
+
+    const getFreshToken = () => {
+        tokenPromise = fetchLiveChatToken().then(
+            token => {
+                pendingTokenRequest = null
+                window.sessionStorage.setItem(cacheKey, JSON.stringify(token))
+                cachedToken = token
+                return token
+            }
+        )
+    }
+
+    const hasToken = () => {
+		if (retrievingToken) {
+			return retrievingToken.then(hasToken)
+		}
+
+		return Promise.resolve(!!cachedToken)
+	}
+
+	const invalidate = () => {
+		cachedToken = null
+		retrievingToken = null
+		return storage.removeItem(cacheKey)
+	}
+
+    return {
+		getToken,
+        getFreshToken,
```

Comment:
```suggestion
        	getFreshToken,
```

Reply:
I think that GH messed something here, its Indented properly 😄